agent/util-scripts/pbench-tool-meister-start

```diff
@@ -732,13 +734,81 @@ def main(_prog: str, cli_params: Namespace) -> int:
+        #
+        # If we can connect, they'll tell us the IP address from which they see
+        # us connecting, and we'll use that (by default) as the server address.
+        localhost = LocalRemoteHost()
+        origin_ip = set()
+        template = TemplateSsh(ssh_cmd, shlex.split(ssh_opts), "echo ${SSH_CONNECTION}")
+        for host in tool_group.hostnames.keys():
+            if not localhost.is_local(host):
+                template.start(host)
+
+        for host, params in tool_group.hostnames.items():
+            if not localhost.is_local(host):
+                connection = template.wait(host)
+                logger.debug("Host %s reports connection `%s`", host, connection)
+                if connection.status == 0 and connection.stdout:
+                    # The SSH_CONNECTION value is the full `stdout` and we
+                    # don't expect a stderr on success; the format is
+                    #   "origin_node origin_port local_node local_port"
+                    # we only need the origin node because we know that's an
+                    # IP for our local TDS and Redis host that the remote can
+                    # reach.
+                    origin = connection.stdout.split()[0]
+                    origin_ip.add(origin)
+
+                    # Store the origin reported by this remote in its parameter
+                    # block: we'll program this as the connection address when
+                    # we send the init handshake to that remote later.
+                    params["origin_host"] = origin
+                else:
+                    # If `ssh` fails, we can't orchestrate anything on this
+                    # remote, so terminate with an error.
+                    logger.error("Host %s reports %s", host, connection)
+                    return ReturnCode.REMOTENOTREACHABLE
+
+        # Process the collected origin addresses from our remotes. Ideally we
+        # have only a single entry here, since the current BaseServer init
+        # (and the Bottle WSGI server) doesn't support listening on multiple
+        # addresses; however we're going to just take the first returned origin
+        # address and hope they can all connect.
+        #
+        # NOTE: an alternative would be (in the absence of an explicit
+        # connection spec) to always listen on '0.0.0.0', but this eliminates
+        # IPv6, and we currently lack the mechanism to listen on both '0.0.0.0'
+        # and '::' (the IPv6 equivalent, aka '0:0:0:0:0:0:0:0').
+        #
+        # NOTE: If the caller/human supplied explicit PBENCH_REDIS_SERVER or
+        # PBENCH_TOOL_DATA_SINK (--redis-server or --tool-data-sink), we'll use
+        # those, but by default we'll use the ssh connection origin instead of
+        # the local `hostname` which may not be reachable by the remotes (or
+        # necessarily routable at all).
+        if len(origin_ip) > 0:
+            logger.debug("Our connection hosts are %s", origin_ip)
```

Comment:
Perhaps make this an informational message when the length of `origin_ip` is greater than one?  Having this is a `debug` message might get lost.

Reply:
I briefly debated a `warning` since it may indicate that some remotes might not be able to reach Redis/TDS. (Or at least, if they can't, this message may provide a clue as to why.) I'm not quite sure why I just left it at `debug`.

Reply:
I'm going to issue a warning if the number of origin IPs isn't exactly 1, because either 0 or > 1 might indicate a routing problem (that is, it's possible that not all remotes will be able to reach our TDS/Redis). I'm going to leave the debug for > 0 just as incidental information on the discovery process.